src/dune_lang/package_info.ml

```diff
@@ -109,6 +109,70 @@ let encode_fields
     ]
 ;;
 
+let valid_maintenance_intent =
+  let open Decoder in
+  map_validate (located string) ~f:(fun (loc, str) ->
+    let rec loop i =
+      if i >= String.length str
+      then Ok ()
+      else (
+        match str.[i] with
+        | '.' -> after_dot (i + 1)
+        | '(' | ')' -> Error "unexpected parenthesis"
+        | _ -> loop (i + 1))
+    and after_dot i =
+      if i >= String.length str
+      then Error "version ends with a dot"
+      else (
+        match str.[i] with
+        | '(' -> parse_token (i + 1) (i + 1)
+        | '.' -> Error "unexpected dot"
+        | _ -> loop (i + 1))
+    and parse_token start i =
+      if i >= String.length str
+      then Error "unclosed parenthesis"
+      else (
+        match str.[i] with
+        | ')' ->
+          let token = String.sub str ~pos:start ~len:(i - start) in
+          if List.mem ~equal:String.equal [ "any"; "latest"; "none" ] token
+          then after_token (i + 1)
+          else
+            Error (Printf.sprintf "unknown intent %S, expected any, latest or none" token)
+        | '-' ->
+          let token = String.sub str ~pos:start ~len:(i - start) in
+          if String.equal token "latest"
+          then parse_num (i + 1) (i + 1)
+          else Error (Printf.sprintf "substraction only allowed for latest, not %S" token)
+        | _ -> parse_token start (i + 1))
+    and parse_num start i =
+      if i >= String.length str
+      then Error ""
+      else (
+        match str.[i] with
+        | ')' when i > start -> after_token (i + 1)
+        | '0' when i > start -> parse_num start (i + 1)
+        | '0' -> parse_num (i + 1) (i + 1)
+        | '1' .. '9' -> parse_num start (i + 1)
+        | _ -> Error "invalid substraction")
+    and after_token i =
+      if i >= String.length str
```

Comment:
Do I understand it right that `""` would be considered valid?

Reply:
I've added a test to show that `""` is an error (and clarified the error message for that special case)